writer.go

```diff
@@ -3,64 +3,182 @@ package log
 import (
 	"io"
 	"sync"
-	"time"
 )
 
+var log = Logger("eventlog")
+
 type MirrorWriter struct {
+	active   bool
+	activelk sync.Mutex
+
+	// channel for incoming writers
+	writerAdd chan io.WriteCloser
+
+	// slices of output writers and their sync channels
 	writers []io.WriteCloser
-	lk      sync.Mutex
+	brChans []chan []byte
+
+	// synchronization channel for incoming writes
+	msgSync chan []byte
+
+	// channel for dead writers to notify the MirrorWriter
+	deadWriter chan io.WriteCloser
+}
+
+func NewMirrorWriter() *MirrorWriter {
+	mw := &MirrorWriter{
+		msgSync:    make(chan []byte, 64), // sufficiently large buffer to avoid callers waiting
+		writerAdd:  make(chan io.WriteCloser),
+		deadWriter: make(chan io.WriteCloser),
+	}
+
+	go mw.logRoutine()
+
+	return mw
 }
 
 func (mw *MirrorWriter) Write(b []byte) (int, error) {
-	mw.lk.Lock()
-	// write to all writers, and nil out the broken ones.
-	var dropped bool
-	done := make(chan error, 1)
-	for i, w := range mw.writers {
-		go func(out chan error) {
+	mw.msgSync <- b
+	return len(b), nil
+}
+
+func (mw *MirrorWriter) handleWriter(w io.WriteCloser, msgs <-chan []byte) {
+	bufLimit := 4096
```

Comment:
it may be good to keep the limit in terms of bytes because some messages may be very big and hose the host.